---
title: "Galaxy Plots on Sims"
author: "Katie Lotterhos"
date: "1/16/2017"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

if(!("ggplot2" %in% installed.packages())){install.packages("ggplot2")}
if(!("gstat" %in% installed.packages())){install.packages("gstat")}
if(!("sp" %in% installed.packages())){install.packages("sp")}   
if(!("maptools" %in% installed.packages())){install.packages("maptools")} 
if(!("ncf" %in% installed.packages())){install.packages("ncf")}
if(!("fields" %in% installed.packages())){install.packages("fields")}  

library(ggplot2)
library(gstat)
library(sp)
library(maptools)
library(ncf)
library(fields)
```

## Evaluating Galaxy plots with known simulation data

Recently I have been exploring the idea of visualizing SNP effect sizes to multiple environmental variables. Here, I am extending the results from the I am using the Lotterhos and Whitlock (2015) simulations to explore this concept in a known scenario. Basically, we want to know if the two distinct groups of behavior we observed in the dataset could be generated by chance if there is only one environmental variable driving the adapatation.

I have chosen three replicate datasets from the LW2015 simulations, that were simulated range expansions in a 1 refugia model, with 90 populations sampled and 20 individuals per population. This demography is similar to the range expansion we expect in lodgepole pine, which is the species and dataset I am applying the concept to (Adaptree data). The total sample size is also similar - in the paper I think we have ~600 diploids (1200 chromosomes), whereas here we have 1800 haploid chromosomes.

In LW2015, we simulated adaptation of loci to a single environmental variable (E_main).

Here, I am going to overlay the Adaptree environmental variables. With this data, I am going to essentially repeat the steps that I took for the Adaptree data:

1) calculate uncorrected Spearman's rho association between allele frequency ($f$) and environment ($E_i$).

2) choose outliers based on uncorrected Spearman's rho as those that have P-values less than the Bonferroni correction. Hereafter these will be called the "candidate set".

3) use Galaxy plots to visualize the candidate set in 2 dimensional space.

I will conduct this exercise for datasets with just neutral loci (9900) and for datasets with neutral loci and loci simulated under varying strengths of selection (9900 and 100).


### Load data

```{r}
print(getwd())
sample_locs <- read.table("data/LW2015samplingdesigns/SchemeRandom1.txt")

dim(sample_locs)
head(sample_locs)
## These were created on a 360 x 360 grid
  range(sample_locs$X_Pops)
  range(sample_locs$Y_Pops)

### Adaptree environments
env_AT <- read.csv("data/MAT06-BLUEs&Climate-pine.csv")
head(env_AT)
which(names(env_AT)=="Latitude")
env_AT <- env_AT[complete.cases(env_AT),20:ncol(env_AT)]

plot(env_AT$Longitude, env_AT$Latitude)

### Transform X and Y to match (0 to 360) while preserving
X1 <- range(env_AT$Longitude, na.rm=TRUE)
Y_range <- range(env_AT$Latitude, na.rm=TRUE)

X1 <- env_AT$Longitude - min(env_AT$Longitude, na.rm=TRUE)
Y1 <- env_AT$Latitude - min(env_AT$Latitude, na.rm=TRUE)
plot(X1, Y1)

# multiply both sides by a factor to get the landscape to fit on 360 x 360

X2 <- X1*360/max(X1, na.rm=TRUE) 
Y2 <- Y1*360/max(Y1, na.rm=TRUE) 

plot(X2, Y2)
```

### Interpolate

For each variable, also compare (i) distributions and (ii) autocorrelation between real environments and resampled environment from the 1R model.

```{r}
env_AT$x <- X2
env_AT$y <- Y2
env_AT2 <- env_AT
coordinates(env_AT) = ~x + y

grd <- expand.grid(x = 1:360, y = 1:360)  # expand points to grid
coordinates(grd) <- ~x + y
gridded(grd) <- TRUE
par(mfrow=c(1,1))
plot(grd, cex = 1.5, col = "grey")
points(env_AT, pch = 1, col = "red", cex = 1)
points(sample_locs$X_Pops, sample_locs$Y_Pops, col="blue", pch="*")

names(env_AT)

for (j in 3:22){
  # interpolate!
  thisvarname <- names(env_AT2)[j]
  print(c(j, thisvarname))
  thisdat <- env_AT[,j]
  names(thisdat) = "e"
  head(thisdat)
  idw <- idw(formula = e ~ 1, locations = thisdat, newdata = grd)  

  idw.output = as.data.frame(idw)  # output is defined as a data table

  names(idw.output)[1:3] <- c("long", "lat", "e.pred")  # give names to the modelled variables

  head(idw.output)
  ggplot() + geom_tile(data = idw.output, aes(x = long, y = lat, fill = e.pred)) + geom_point(data = env_AT2, aes(x = X2, y = Y2), shape = 21, 
        colour = "red") + geom_point(data = sample_locs, aes(x = X_Pops, y = Y_Pops), shape = 25,
        colour = "yellow")

  #head(sample_locs)

  for (i in 1:nrow(sample_locs)){
    wrow <- which(idw.output$long==sample_locs$X_Pops[i] & idw.output$lat==sample_locs$Y_Pops[i])
    sample_locs$e[i] <- idw.output$e.pred[wrow]
  }
  head(sample_locs)
  names(sample_locs)[ncol(sample_locs)] <- thisvarname

  par(mfrow=c(2,1), mar=c(2,4,2,0))
  b<-hist(env_AT2[,j], main=paste("Real", thisvarname))
  hist(sample_locs[,ncol(sample_locs)], breaks=b$breaks, main="Interpolated")


  ncf.cor.real <- correlog(env_AT2$x, env_AT2$y, env_AT2[,j],  increment=2, resamp=500, quiet = TRUE)

  ncf.cor.sample <- correlog(sample_locs$X_Pops, sample_locs$Y_Pops, sample_locs[,ncol(sample_locs)],  increment=2, resamp=500, quiet = TRUE)

  par(mfrow=c(2,1), mar=c(2,4,2,0))
  plot(ncf.cor.real$correlation, ylim= c(-2,2), main=paste("True",thisvarname), type="l")
  plot(ncf.cor.sample$correlation, ylim= c(-2,2), main="Interpolated", type="l")

}
```


# Compare correlation matrices of raw and interpolated environments 

Compare correlation structure between real environments and resampled environment from the 1R model.

```{r}
head(env_AT2)
(creal <- cor(env_AT2[,1:22]))
head(sample_locs)
names(sample_locs)
(csim <- cor(sample_locs[,c(2:3,8:27)]))

plot(creal, csim)
abline(0,1)


par(mfrow=c(2,1))
col <- two.colors(start="red", end="blue", middle="grey")
image.plot(creal, col=col)
image.plot(csim, col=col)
```


### Write simulated environments to file
```{r}
write.table(sample_locs, "data/results_AdaptreeEnviFor_R90.txt")
```